### PR TITLE
Add telemetry filters

### PR DESCRIPTION
**Pull request by WayneFerrao**
[AB#3871](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3871)

## Description
Added filters for the event category and event name for telemetry events using Fluent UI components.

imageng)

### CODE REVIEW

File: packages/tools/client-debugger/client-debugger-view/src/components/ContainerDevtoolsView.tsx [hunk 1]
```diff
@@ -2,7 +2,8 @@
  * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
  * Licensed under the MIT License.
  */
-import { IOverflowSetItemProps, IconButton, Link, OverflowSet, Stack } from "@fluentui/react";
+import { Stack } from "@fluentui/react";
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/ContainerDevtoolsView.tsx:5]: Nit: The container data view changes seem like they could be their own PR.

File: packages/tools/client-debugger/client-debugger-view/src/components/ContainerDevtoolsView.tsx [hunk 2]
```diff
@@ -155,10 +160,17 @@ function _ContainerDevtoolsView(props: _ContainerDevtoolsViewProps): React.React
 			</Stack.Item>
 			<Stack.Item style={{ width: "100%", height: "100%", overflowY: "auto" }}>
 				<Stack tokens={{ childrenGap: 10 }}>
-					<PanelViewSelectionMenu
-						currentSelection={innerViewSelection}
-						updateSelection={setInnerViewSelection}
-					/>
+					<TabList selectedValue={innerViewSelection} onTabSelect={onTabSelect}>
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/ContainerDevtoolsView.tsx:163]: The previous implementation was able to generate new tabs dynamically if we added a new flag to the `PanelView` enum. Could we update this logic to preserve that? I.e., generate the list of tab options from the enum flags?

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 3]
```diff
@@ -12,13 +12,18 @@ import {
...
+	Combobox,
+	ComboboxProps,
 	ToggleButton,
 	DataGridBody,
 	DataGridRow,
 	DataGrid,
 	DataGridHeader,
 	DataGridHeaderCell,
 	DataGridCell,
+	Dropdown as DropDown,
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:24]: Is `as DropDown` needed?

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 4]
```diff
@@ -64,6 +69,37 @@ export function TelemetryView(): React.ReactElement {
 		ITimestampedTelemetryEvent[] | undefined
 	>();
 	const [maxEventsToDisplay, setMaxEventsToDisplay] = React.useState<number>(DEFAULT_PAGE_SIZE);
+	const [selectedCategory, setSelectedCategory] = useState("");
+	const [filteredTelemetryEvents, setFilteredTelemetryEvents] = React.useState<
+		ITimestampedTelemetryEvent[] | undefined
+	>();
+	const [customSearch, setCustomSearch] = React.useState("");
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:76]: It would be good to document the semantics of each of these pieces of search / filter state. In particular how they are used (exact matches? partial matches? What does empty string / array mean? etc.).

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 5]
```diff
@@ -64,6 +69,37 @@ export function TelemetryView(): React.ReactElement {
...
+	const [eventNameOptions, setEventNameOptions] = useState<string[]>([]);
+	const [matchingOptions, setMatchingOptions] = React.useState([...eventNameOptions]);
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:78]: Initializing your state to another piece of state likely won't work out how you want. Probably best to initialize this to empty or undefined, and update it alongside the other state within the effect.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 6]
```diff
@@ -64,6 +69,37 @@ export function TelemetryView(): React.ReactElement {
...
+
+	React.useEffect(() => {
+		// Create list of all event names
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:81]: I would highly recommend extracting all of your filtering logic out into a separate function, and call that from the component / effect as needed. That way you can write and document the logic strictly in terms of inputs (unfiltered data + filter options) and outputs (filtered data). We can even unit test the logic that way if it becomes complex enough.

**Review by Josmithr — COMMENTED** [review]
I love the visual changes. I left some feedback about code organization and state management.

I also think it would be worth splitting this into 2 separate PRs: 1 for the container data tab changes, and one for the telemetry view changes.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx (hunk 3 above)
Comment (WayneFerrao) [reply packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:24]: Yes, I added this because there's a Dropdown component from @fluentui/react and @fluentui/react-components. I wanted to keep the filter components consistent. 
To keep the entire UI consistent, I can update the max events displayed dropdown to also be from @fluentui/react-components

Comment (alexvy86) [reply packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:24]: Yeah, I'd say let's clean up that now and use the same dropdown everywhere.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 7]
```diff
@@ -155,7 +236,11 @@ export function TelemetryView(): React.ReactElement {
 		createTableColumn<Item>({
 			columnId: "information",
 			renderHeaderCell: () => {
-				return <b>Information</b>;
+				return (
+					<div>
+						<h2>Information</h2>
+					</div>
```
Comment (seanimam) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:242]: Is there a need for an outer div here?

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 8]
```diff
@@ -180,6 +265,60 @@ export function TelemetryView(): React.ReactElement {
...
+
+	const handleCategoryChange: DropdownProps["onOptionSelect"] = (event, data) => {
+		const category = data.optionText !== undefined ? data.optionText : "";
+		setSelectedCategory(category);
+	};
+
+	/**
+	 * Event handler that updates table to display events matching text in event name filter.
+	 */
+	const onChange: ComboboxProps["onChange"] = (event) => {
```
Comment (seanimam) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:294]: Rename this function from onChange to specify what is changing

Comment (seanimam) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:294]: Rename this function from onChange to specify what is changing

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx (hunk 5 above)
Comment (seanimam) [reply packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:78]: Yep, The current code would initialize the default value of matchingOptions to eventNameOptions (which is initialized to an empty array).

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 9]
```diff
@@ -64,6 +69,37 @@ export function TelemetryView(): React.ReactElement {
...
+		]);
+		// Initialy matching options are all options
+		setMatchingOptions(eventNameOptions.sort());
+		const filteredEvents = telemetryEvents?.filter((event) => {
+			return (
+				(selectedCategory === "" ||
+					selectedCategory === "All" ||
+					event.logContent.category === selectedCategory) &&
+				(customSearch === "" ||
+					event.logContent.eventName.slice("fluid:telemetry:".length) === customSearch)
```
Comment (seanimam) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:97]: Personal preference, but I think it's worth splitting this from one large chained set boolean operations into some if else statements. It'll be much easier to understand and maintain it in the future.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx (hunk 4 above)
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:76]: Also, if there is any state that we are tracking here that is strictly derived from other state and input props, we probably don't need to store it as state (we will have to recompute it any time the other state / props change anyways, so we don't really need to store it). Not sure if that's the case here or not, but figured I would call it out as a best practice,

**Comment by jikim-msft** [comment]
NIT

Do you think you can give a bit of margin to the header of the telemetry? It looks more aligned with the margin. 

MicrosoftTeams-image

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 10]
```diff
@@ -99,6 +124,28 @@ export function TelemetryView(): React.ReactElement {
 		};
 	}, [messageRelay, setTelemetryEvents]);
 
+	/**
+	 * Filters all telemetry events based on category and event name
+	 * @returns filtered list of events
+	 */
+	function getFilteredEvents(): ITimestampedTelemetryEvent[] | undefined {
+		let filteredEvents = telemetryEvents;
```
Comment (jikim-msft) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:132]: Is this line of code necessary?

Comment (WayneFerrao) [reply packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:132]: Yes I initialize filteredEvents to all telemetryEvents. This would represent the initial state when the user has not set any filters.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx (hunk 4 above)
Comment (Josmithr) [reply packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:73]: Re-opening, because I think the other pieces of state could use docs too. `eventNameOptions` and `matchingOptions`, for example, aren't completely evident based on their variable names. It would be nice to document how they are used, and what default states (i.e. `[]`) mean.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 11]
```diff
@@ -64,6 +62,33 @@ export function TelemetryView(): React.ReactElement {
...
+	>();
+	/**
+	 * Used to store query for the searchable dropdown. The query is used to perform
+	 * partial match searches and will display all events if query is an empty string.
+	 */
+	const [customSearch, setCustomSearch] = React.useState("");
+	const [eventNameOptions, setEventNameOptions] = useState<string[]>([]);
+	const [matchingOptions, setMatchingOptions] = React.useState<string[]>([]);
+
+	React.useEffect(() => {
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:77]: It might be a good idea to extract all of the filtering logic into a separate component. Something like "FilteredTelemetryView". Then we could have the following breakdown (or something like it):
* TelemetryView: Does the message-passing-based communication with the webpage to get the unfiltered events and store them.
* FilteredTelemetryView: Takes in the unfiltered list of telemetry events, and applies filters to them.

Recommended best practice is to make components "single responsibility". This can be a bit nebulous, but aiming to separate responsibilities can definitely be helpful for detangling complex state management.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 12]
```diff
@@ -259,32 +405,41 @@ interface _ListLengthSelectionProps {
  */
 function _ListLengthSelection(props: _ListLengthSelectionProps): React.ReactElement {
 	const { currentLimit, onChangeSelection } = props;
-	const dropdownStyles: Partial<IDropdownStyles> = {
-		dropdown: { width: "300px", zIndex: "1" },
-	};
-
+	// const [eventsDisplayed, setEventsDisplayed] = useState(50);
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:408]: Should this be removed?

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 13]
```diff
@@ -124,6 +98,121 @@ export function TelemetryView(): React.ReactElement {
...
+				</Dropdown>
+			</div>
+		</Stack>
+	);
+}
+
+/**
+ * {@link _FilteredTelemetryViewProps} input props.
+ */
+interface _FilteredTelemetryViewProps {
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:179]: Can we remove the _ from this?

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 14]
```diff
@@ -326,121 +416,47 @@ export function TelemetryView(): React.ReactElement {
...
-					})}
-				</Dropdown>
-			</div>
-		</Stack>
+					)}
+				</DataGridBody>
+			</DataGrid>
+		</>
+	) : (
+		<Waiting label={"Waiting for Telemetry events"} />
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:460]: The Waiting handling should probably go in the parent component, since that's the one that is communicating with the app. This component shouldn't have to worry about the `undefined` case.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 15]
```diff
@@ -99,6 +98,143 @@ export function TelemetryView(): React.ReactElement {
...
+				<FilteredTelemetryView telemetryEvents={telemetryEvents} />
+			</StackItem>
+		</Stack>
+	);
+}
+
+/**
+ * {@link _ListLengthSelectionProps} input props.
+ */
+interface _ListLengthSelectionProps {
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:119]: I don't think these need to be prefixed with "_". In other places we did that to differentiate components in the same module that would otherwise have the same name, but we generally don't use "_" for private members.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 16]
```diff
@@ -99,6 +98,143 @@ export function TelemetryView(): React.ReactElement {
...
+			</StackItem>
+			<StackItem>
+				<FilteredTelemetryView telemetryEvents={telemetryEvents} />
+			</StackItem>
+		</Stack>
+	);
+}
+
+/**
+ * {@link _ListLengthSelectionProps} input props.
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:117]:
```suggestion
 * {@link _ListLengthSelection} input props.
```

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 17]
```diff
@@ -99,6 +98,143 @@ export function TelemetryView(): React.ReactElement {
...
+						);
+					})}
+				</Dropdown>
+			</div>
+		</Stack>
+	);
+}
+
+/**
+ * {@link _FilteredTelemetryViewProps} input props.
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:177]:
```suggestion
 * {@link FilteredTelemetryView} input props.
```

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx (hunk 15 above)
Comment (alexvy86) [reply packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:119]: Markdown strikes again 😄. There's supposed to be a `_` between each pair of quotes.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 18]
```diff
@@ -99,6 +98,147 @@ export function TelemetryView(): React.ReactElement {
...
+	);
+}
+
+/**
+ * {@link ListLengthSelection} input props.
+ */
+interface ListLengthSelectionProps {
+	/**
+	 * The current limit (max number of telemetry events to show).
+	 * @defaultValue {@link DEFAULT_PAGE_SIZE}
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:126]: Nit: This property isn't optional, so I don't think the defaultValue comment here adds value. This is policy set by the consumer. Generally `@defaultValue` accompanies optional properties.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 19]
```diff
@@ -99,6 +98,147 @@ export function TelemetryView(): React.ReactElement {
...
+		</Stack>
+	);
+}
+
+/**
+ * {@link FilteredTelemetryView} input props.
+ */
+interface FilteredTelemetryViewProps {
+	/**
+	 * The current limit (max number of telemetry events to show).
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:185]: I think these must have been copy-pasted from something else; I don't think these are right.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 20]
```diff
@@ -99,6 +98,147 @@ export function TelemetryView(): React.ReactElement {
...
+	 * @defaultValue {@link DEFAULT_PAGE_SIZE}
+	 */
+	telemetryEvents: ITimestampedTelemetryEvent[] | undefined;
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:188]: Does this need to allow `undefined`? Since the parent component handles the Waiting case, I think this can just be:
```suggestion
	telemetryEvents: ITimestampedTelemetryEvent[];
```

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 21]
```diff
@@ -99,6 +98,147 @@ export function TelemetryView(): React.ReactElement {
...
+}
+
+function FilteredTelemetryView(props: FilteredTelemetryViewProps): React.ReactElement {
+	const { telemetryEvents } = props;
+	const [eventNameOptions, setEventNameOptions] = useState<string[]>([]);
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:193]: I think we should document each of these pieces of state to explain what they are, and what their default values mean (undefined, empty array, etc.). This component's state is pretty complex, so clear documentation will make future maintenance much easier.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 22]
```diff
@@ -99,6 +98,147 @@ export function TelemetryView(): React.ReactElement {
...
+		setMatchingOptions(eventNameOptions.sort());
+		const filteredEvents = getFilteredEvents();
+		setFilteredTelemetryEvents(filteredEvents);
+	}, [telemetryEvents, selectedCategory, customSearch]);
+
+	/**
+	 * Filters all telemetry events based on category and event name
+	 * @returns filtered list of events
+	 */
+	function getFilteredEvents(): ITimestampedTelemetryEvent[] | undefined {
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:224]: Nit: Probably best to move this inside of your `useEffect` block. Otherwise we would need to add it as a dependency in the effect dependencies list.

File: packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx [hunk 23]
```diff
@@ -180,112 +366,99 @@ export function TelemetryView(): React.ReactElement {
...
-				</DataGrid>
-			</>
-		) : (
-			<Waiting label={"Waiting for Telemetry events"} />
-		);
+	/**
+	 * Gets list of valid categories for displayed telemetry events.
+	 * @returns list of option
+	 */
+	function getCategories(): { key: string; text: string }[] {
```
Comment (Josmithr) [inline packages/tools/client-debugger/client-debugger-view/src/components/TelemetryView.tsx:373]: Nit: I would recommend moving this to be above the rendering code that calls it.

**Review by Josmithr — APPROVED** [review]
I left a few more comments, mostly about code organization and documentation. Otherwise, I think this looks great!
